### ISSUE TITLE
Invalid route POST request results in key space error 500 instead of 404

### ISSUE THREAD

**Issue by valerauko** [rails/rails#51056]
### Steps to reproduce

It is possible to craft a POST request that results in a "exceeded available parameter key space" error, even though the request is targeting a route that doesn't exist. We've seen this error occur in production (due to malicious requests), but we don't actually have such a crafted POST request at hand to reproduce.

#### Config

We don't have Rack::Utils.key_space_limit configured, so it should be the default value.
```
pry(main)> Rack::Utils.key_space_limit
=> 65536
```

### Expected behavior

a 404 error is returned

### Actual behavior

an 500 internal server error is returned

```
Invalid or incomplete POST params
Error during failsafe response: exceeded available parameter key space
  /usr/local/bundle/gems/rack-2.2.8/lib/rack/query_parser.rb:175:in `[]='
  /usr/local/bundle/gems/rack-2.2.8/lib/rack/query_parser.rb:103:in `normalize_params'
  /usr/local/bundle/gems/rack-2.2.8/lib/rack/query_parser.rb:75:in `block in parse_nested_query'
  /usr/local/bundle/gems/rack-2.2.8/lib/rack/query_parser.rb:72:in `each'
  /usr/local/bundle/gems/rack-2.2.8/lib/rack/query_parser.rb:72:in `parse_nested_query'
  /usr/local/bundle/gems/rack-2.2.8/lib/rack/request.rb:590:in `parse_query'
  /usr/local/bundle/gems/rack-2.2.8/lib/rack/request.rb:454:in `POST'
  /usr/local/bundle/gems/actionpack-6.1.7.6/lib/action_dispatch/http/request.rb:400:in `block (2 levels) in POST'
  /usr/local/bundle/gems/actionpack-6.1.7.6/lib/action_dispatch/http/parameters.rb:90:in `block in parse_formatted_parameters'
  /usr/local/bundle/gems/actionpack-6.1.7.6/lib/action_dispatch/http/parameters.rb:90:in `fetch'
  /usr/local/bundle/gems/actionpack-6.1.7.6/lib/action_dispatch/http/parameters.rb:90:in `parse_formatted_parameters'
  /usr/local/bundle/gems/actionpack-6.1.7.6/lib/action_dispatch/http/request.rb:399:in `block in POST'
  /usr/local/bundle/gems/rack-2.2.8/lib/rack/request.rb:69:in `fetch'
  /usr/local/bundle/gems/rack-2.2.8/lib/rack/request.rb:69:in `fetch_header'
  /usr/local/bundle/gems/actionpack-6.1.7.6/lib/action_dispatch/http/request.rb:398:in `POST'
  /usr/local/bundle/gems/actionpack-6.1.7.6/lib/action_dispatch/http/parameters.rb:55:in `parameters'
  /usr/local/bundle/gems/actionpack-6.1.7.6/lib/action_dispatch/http/mime_negotiation.rb:166:in `params_readable?'
  /usr/local/bundle/gems/actionpack-6.1.7.6/lib/action_dispatch/http/mime_negotiation.rb:71:in `block in formats'
  /usr/local/bundle/gems/rack-2.2.8/lib/rack/request.rb:69:in `fetch'
  /usr/local/bundle/gems/rack-2.2.8/lib/rack/request.rb:69:in `fetch_header'
  /usr/local/bundle/gems/actionpack-6.1.7.6/lib/action_dispatch/http/mime_negotiation.rb:70:in `formats'
  /usr/local/bundle/gems/actionpack-6.1.7.6/lib/action_dispatch/middleware/show_exceptions.rb:64:in `fallback_to_html_format_if_invalid_mime_type'
  /usr/local/bundle/gems/actionpack-6.1.7.6/lib/action_dispatch/middleware/show_exceptions.rb:50:in `render_exception'
  /usr/local/bundle/gems/actionpack-6.1.7.6/lib/action_dispatch/middleware/show_exceptions.rb:36:in `rescue in call'
  /usr/local/bundle/gems/actionpack-6.1.7.6/lib/action_dispatch/middleware/show_exceptions.rb:31:in `call'
  /usr/local/bundle/gems/railties-6.1.7.6/lib/rails/rack/logger.rb:37:in `call_app'
  /usr/local/bundle/gems/railties-6.1.7.6/lib/rails/rack/logger.rb:26:in `block in call'
  /usr/local/bundle/gems/activesupport-6.1.7.6/lib/active_support/tagged_logging.rb:99:in `block in tagged'
  /usr/local/bundle/gems/activesupport-6.1.7.6/lib/active_support/tagged_logging.rb:37:in `tagged'
  /usr/local/bundle/gems/activesupport-6.1.7.6/lib/active_support/tagged_logging.rb:99:in `tagged'
  /usr/local/bundle/gems/railties-6.1.7.6/lib/rails/rack/logger.rb:26:in `call'
  /usr/local/bundle/gems/actionpack-6.1.7.6/lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
  /usr/local/bundle/gems/actionpack-6.1.7.6/lib/action_dispatch/middleware/request_id.rb:26:in `call'
  /usr/local/bundle/gems/rack-2.2.8/lib/rack/method_override.rb:24:in `call'
  /usr/local/bundle/gems/rack-2.2.8/lib/rack/runtime.rb:22:in `call'
  /usr/local/bundle/gems/activesupport-6.1.7.6/lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
  /usr/local/bundle/gems/actionpack-6.1.7.6/lib/action_dispatch/middleware/executor.rb:14:in `call'
  /usr/local/bundle/gems/actionpack-6.1.7.6/lib/action_dispatch/middleware/static.rb:24:in `call'
  /usr/local/bundle/gems/rack-2.2.8/lib/rack/sendfile.rb:110:in `call'
  /usr/local/bundle/gems/actionpack-6.1.7.6/lib/action_dispatch/middleware/host_authorization.rb:142:in `call'
  /usr/local/bundle/gems/ddtrace-1.12.1/lib/datadog/tracing/contrib/rack/middlewares.rb:98:in `call'
  /usr/local/bundle/gems/railties-6.1.7.6/lib/rails/engine.rb:539:in `call'
  /usr/local/bundle/gems/railties-6.1.7.6/lib/rails/railtie.rb:207:in `public_send'
  /usr/local/bundle/gems/railties-6.1.7.6/lib/rails/railtie.rb:207:in `method_missing'
  /usr/local/bundle/gems/unicorn-5.7.0/lib/unicorn/http_server.rb:632:in `process_client'
  /usr/local/bundle/gems/unicorn-worker-killer-0.4.4/lib/unicorn/worker_killer.rb:92:in `process_client'
  /usr/local/bundle/gems/unicorn-5.7.0/lib/unicorn/http_server.rb:728:in `worker_loop'
  /usr/local/bundle/gems/unicorn-5.7.0/lib/unicorn/http_server.rb:548:in `spawn_missing_workers'
  /usr/local/bundle/gems/unicorn-5.7.0/lib/unicorn/http_server.rb:562:in `maintain_worker_count'
  /usr/local/bundle/gems/unicorn-5.7.0/lib/unicorn/http_server.rb:295:in `join'
  /usr/local/bundle/gems/unicorn-5.7.0/bin/unicorn_rails:209:in `<top (required)>'
  /usr/local/bundle/bin/unicorn_rails:25:in `load'
  /usr/local/bundle/bin/unicorn_rails:25:in `<top (required)>'
  /usr/local/lib/ruby/3.0.0/bundler/cli/exec.rb:58:in `load'
  /usr/local/lib/ruby/3.0.0/bundler/cli/exec.rb:58:in `kernel_load'
  /usr/local/lib/ruby/3.0.0/bundler/cli/exec.rb:23:in `run'
  /usr/local/lib/ruby/3.0.0/bundler/cli.rb:478:in `exec'
  /usr/local/lib/ruby/3.0.0/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
  /usr/local/lib/ruby/3.0.0/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
  /usr/local/lib/ruby/3.0.0/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
  /usr/local/lib/ruby/3.0.0/bundler/cli.rb:31:in `dispatch'
  /usr/local/lib/ruby/3.0.0/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
  /usr/local/lib/ruby/3.0.0/bundler/cli.rb:25:in `start'
  /usr/local/lib/ruby/gems/3.0.0/gems/bundler-2.2.32/libexec/bundle:49:in `block in <top (required)>'
  /usr/local/lib/ruby/3.0.0/bundler/friendly_errors.rb:103:in `with_friendly_errors'
  /usr/local/lib/ruby/gems/3.0.0/gems/bundler-2.2.32/libexec/bundle:37:in `<top (required)>'
  /usr/local/bin/bundle:23:in `load'
  /usr/local/bin/bundle:23:in `<main>'
```

This seems to be related to https://github.com/rails/rails/issues/29983

### System configuration
**Rails version**: 6.1.7.6

**Ruby version**: 3.0.3

**Comment by williantenfen** [comment]
I can see that `.key_space_limit` is deprecated and will be removed in Rack 3.1 (https://github.com/rack/rack/pull/1757)
But there's still one option `.param_depth_limit` which will raise the same exception `ParamsTooDeepError` .

I can see that there is something on action_dispatch doing some errors handling from Rack expection:

https://github.com/rails/rails/blob/057563a888fbb3f9237e52b70915edcaf9d797fb/actionpack/lib/action_dispatch/middleware/exception_wrapper.rb#L26

Would this be the case to add a wrapper for `Rack::QueryParser::ParamsTooDeepError` to return a bad_request response? ( If thats the case, I can issue a small PR)

**Comment by skipkayhil** [comment]
I believe this was fixed in #45997, so you should upgrade Rails to 7+ for the fix.

Closing since it appears to be addressed already, but feel free to reply if you think that is not the case.
